cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkTransactionClient.java

```diff
@@ -0,0 +1,109 @@
+import java.net.URI;
+import java.net.URL;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import javax.annotation.Nullable;
+
+/**
+ * Client class to interact with {@link SparkTransactionService} through HTTP. It is used by tasks execute inside
+ * executor processes.
+ */
+final class SparkTransactionClient {
+
+  private static final TransactionCodec TX_CODEC = new TransactionCodec();
+  private static final long DEFAULT_TX_POLL_INTERVAL_MS = 50;
+
+  private final URI txServiceBaseURI;
+  private final long txPollIntervalMillis;
+
+  SparkTransactionClient(URI txServiceBaseURI) {
+    this(txServiceBaseURI, DEFAULT_TX_POLL_INTERVAL_MS);
+  }
+
+  SparkTransactionClient(URI txServiceBaseURI, long txPollIntervalMillis) {
+    this.txServiceBaseURI = txServiceBaseURI;
+    this.txPollIntervalMillis = txPollIntervalMillis;
+  }
+
+  /**
+   * Returns the {@link Transaction} for the given stage.
+   *
+   * @param stageId the stage id to query for {@link Transaction}.
+   * @param timeout the maximum time to wait
+   * @param timeUnit the time unit of the timeout argument
+   * @return the {@link Transaction} to be used for the given stage.
+   *
+   * @throws TimeoutException if the wait timed out
+   * @throws InterruptedException if the current thread was interrupted while waiting
+   * @throws TransactionFailureException if failed to get transaction for the given stage. Calling this method again
+   *                                     with the same stage id will result in the same exception
+   */
+  Transaction getTransaction(int stageId, long timeout,
+                             TimeUnit timeUnit) throws TimeoutException, InterruptedException,
+                                                       TransactionFailureException {
+    long timeoutMillis = Math.max(0L, timeUnit.toMillis(timeout) - txPollIntervalMillis);
+    Stopwatch stopwatch = new Stopwatch().start();
+    Transaction transaction = getTransaction(stageId);
+
+    while (transaction == null && stopwatch.elapsedMillis() < timeoutMillis) {
+      TimeUnit.MILLISECONDS.sleep(txPollIntervalMillis);
+      transaction = getTransaction(stageId);
+    }
+    if (transaction == null) {
+      throw new TimeoutException("Cannot get transaction for stage " + stageId + " after " + timeout + " " + timeUnit);
+    }
+    return transaction;
+  }
+
+  @Nullable
+  private Transaction getTransaction(int stageId) throws TransactionFailureException {
+    try {
+      URL url = txServiceBaseURI.resolve("/spark/stages/" + stageId + "/transaction").toURL();
+      HttpURLConnection urlConn = (HttpURLConnection) url.openConnection();
+      try {
+        int responseCode = urlConn.getResponseCode();
+        if (responseCode == 200) {
+          return TX_CODEC.decode(ByteStreams.toByteArray(urlConn.getInputStream()));
+        }
+        if (responseCode == 404) {
+          return null;
+        }
+        throw new TransactionFailureException(
```

Comment:
I think it would be good to have an exception hierarchy for transaction failures. When I originally wrote it, there were only two types of failures: a conflict, or an exception from within the tx. They uses different transactions (txConflictException for conflits), so when you catch an exception you can easily tell them apart. Does it make sense to have a subclass for this case?